Compilation readjustment

ferxxp opened the pull request:
Solved a compiling error due to templates and added a Dockerfile for future-proofing compilations.

ferxxp committed:
Template compiler error mitigated

ferxxp committed:
Added Docker for future-proofing

jvgomez commented on CMakeLists.txt:
```diff
@@ -18,6 +18,9 @@ set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "-O3 -g")
 set(CMAKE_CXX_FLAGS_DEBUG  "-Wall -Wno-unused-local-typedefs -g")
 
+#Flag to avoid error in compilation
+add_compile_options(-fpermissive)
```
I am curious, which compile error are you observing?

ferxxp replied:
When I was compiling on a base ubuntu image (the one for dockers), after installing all dependencies and  running cmake, the make command failed with the following console log:

```[ 25%] Built target fast_methods
Consolidate compiler generated dependencies of target fm_benchmark
[ 31%] Building CXX object CMakeFiles/fm_benchmark.dir/src/benchmark/fm_benchmark.cpp.o
In file included from /planning/include/fast_methods/fm/fmm.hpp:53,
                 from /planning/src/benchmark/fm_benchmark.cpp:25:
/planning/include/fast_methods/fm/eikonalsolver.hpp: In member function 'virtual double EikonalSolver<grid_t>::solveEikonal(const int&)':
/planning/include/fast_methods/fm/eikonalsolver.hpp:55:22: error: there are no arguments to 'isinf' that depend on a template parameter, so a declaration of 'isinf' must be available [-fpermissive]
   55 |                 if (!isinf(minTInDim) && minTInDim < grid_->getCell(idx).getArrivalTime())
      |                      ^~~~~
/planning/include/fast_methods/fm/eikonalsolver.hpp:55:22: note: (if you use '-fpermissive', G++ will accept your code, but allowing the use of an undeclared name is deprecated)
In file included from /usr/include/boost/config/header_deprecated.hpp:18,
                 from /usr/include/boost/progress.hpp:22,
                 from /planning/include/fast_methods/benchmark/benchmark.hpp:27,
                 from /planning/src/benchmark/fm_benchmark.cpp:34:
/usr/include/boost/progress.hpp: At global scope:
/usr/include/boost/progress.hpp:23:1: note: '#pragma message: This header is deprecated. Use the facilities in <boost/timer/timer.hpp> or <boost/timer/progress_display.hpp> instead.'
   23 | BOOST_HEADER_DEPRECATED( "the facilities in <boost/timer/timer.hpp> or <boost/timer/progress_display.hpp>" )
      | ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/boost/timer.hpp:21:1: note: '#pragma message: This header is deprecated. Use the facilities in <boost/timer/timer.hpp> instead.'
   21 | BOOST_HEADER_DEPRECATED( "the facilities in <boost/timer/timer.hpp>" )
      | ^~~~~~~~~~~~~~~~~~~~~~~
In file included from /planning/include/fast_methods/fm/fmm.hpp:53,
                 from /planning/src/benchmark/fm_benchmark.cpp:25:
/planning/include/fast_methods/fm/eikonalsolver.hpp: In instantiation of 'double EikonalSolver<grid_t>::solveEikonal(const int&) [with grid_t = nDGridMap<FMCell, 3>]':
/planning/include/fast_methods/fm/eikonalsolver.hpp:48:24:   required from here
/planning/include/fast_methods/fm/eikonalsolver.hpp:55:27: error: 'isinf' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
   55 |                 if (!isinf(minTInDim) && minTInDim < grid_->getCell(idx).getArrivalTime())
      |                      ~~~~~^~~~~~~~~~~
In file included from /planning/src/benchmark/fm_benchmark.cpp:16:
/usr/include/c++/11/cmath:584:3: note: 'constexpr bool std::isinf(float)' declared here, later in the translation unit
  584 |   isinf(float __x)
      |   ^~~~~
In file included from /planning/include/fast_methods/fm/fmm.hpp:53,
                 from /planning/src/benchmark/fm_benchmark.cpp:25:
/planning/include/fast_methods/fm/eikonalsolver.hpp: In instantiation of 'double EikonalSolver<grid_t>::solveEikonal(const int&) [with grid_t = nDGridMap<FMCell, 2>]':
/planning/include/fast_methods/fm/eikonalsolver.hpp:48:24:   required from here
/planning/include/fast_methods/fm/eikonalsolver.hpp:55:27: error: 'isinf' was not declared in this scope, and no declarations were found by argument-dependent lookup at the point of instantiation [-fpermissive]
   55 |                 if (!isinf(minTInDim) && minTInDim < grid_->getCell(idx).getArrivalTime())
      |                      ~~~~~^~~~~~~~~~~
In file included from /planning/src/benchmark/fm_benchmark.cpp:16:
/usr/include/c++/11/cmath:584:3: note: 'constexpr bool std::isinf(float)' declared here, later in the translation unit
  584 |   isinf(float __x)
      |   ^~~~~
make[2]: *** [CMakeFiles/fm_benchmark.dir/build.make:76: CMakeFiles/fm_benchmark.dir/src/benchmark/fm_benchmark.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:137: CMakeFiles/fm_benchmark.dir/all] Error 2
make: *** [Makefile:136: all] Error 2 
```


As stated in the error, adding

```
add_compile_options(-fpermissive)
```
solves the error and compiles it successfully.

Just if you want to check  I used the following commands:

```
docker run -it --rm ubuntu

apt-get update
apt-get install build-essential -y
apt-get install nano cmake git doxygen cmake-curses-gui wget -y
apt-get install imagemagick libboost-all-dev cimg-dev  #requires user input

mkdir /planning
git clone https://github.com/jvgomez/fast_methods.git /planning
cd planning
mkdir build
cmake -S /planning/ -B /planning/build/
cd build
make
```

jvgomez replied:
Thanks for the info.  I am a bit worried about adding -fpermissive as it mostly hides the error, and doesn't solve it (https://stackoverflow.com/questions/8843818/what-does-the-fpermissive-flag-do).

Could you please try replacing all the isinf with std::isinf instead?

jvgomez commented on Dockerfile:
```diff
@@ -0,0 +1,17 @@
+ARG DEBIAN_FRONTEND=noninteractive #should be change
+RUN apt-get install imagemagick libboost-all-dev cimg-dev -y
+
+RUN git clone https://github.com/ferxxp/fast_methods_2023.git /fast_methods/
```
This is a great contribution, thank you!
Since this will be uploaded to this repo, perhaps we should use the git repo jvgomez/fast_methods ?

ferxxp replied:
Thanks.

I'm a little noob with git and github, but yes. 
If you accept the merge I believe it will be added to the repo jvgomez/fast_methods.

jvgomez replied:
What I meant is that this dockerfile should clone this repo (otherwise it will clone a fork)
```suggestion
RUN git clone https://github.com/jvgomez/fast_methods.git /fast_methods/

```

ferxxp committed:
Update Dockerfile

Co-authored-by: Javier V. Gómez <javvgomez@gmail.com>

ferxxp committed:
testing compiler

ferxxp committed:
typo

ferxxp committed:
isinf replaced by std::isinf

ferxxp committed:
isnan replace with std::isnan

ferxxp committed:
testing solutions for compilation error

just trying

?

ferxxp committed:
headder problem?

ferxxp reviewed (commented):
This addition is t sove the following compilation error:

````
[  6%] Building CXX object CMakeFiles/fast_methods.dir/src/console/console.cpp.o
[ 12%] Building CXX object CMakeFiles/fast_methods.dir/src/ndgridmap/cell.cpp.o
[ 18%] Building CXX object CMakeFiles/fast_methods.dir/src/ndgridmap/fmcell.cpp.o
[ 25%] Linking CXX shared library libfast_methods.so
[ 25%] Built target fast_methods
[ 31%] Building CXX object CMakeFiles/fm_benchmark.dir/src/benchmark/fm_benchmark.cpp.o
In file included from /usr/include/boost/config/header_deprecated.hpp:18,
                 from /usr/include/boost/progress.hpp:22,
                 from /planning/include/fast_methods/benchmark/benchmark.hpp:27,
                 from /planning/src/benchmark/fm_benchmark.cpp:34:
/usr/include/boost/progress.hpp:23:1: note: '#pragma message: This header is deprecated. Use the facilities in <boost/timer/timer.hpp> or <boost/timer/progress_display.hpp> instead.'
   23 | BOOST_HEADER_DEPRECATED( "the facilities in <boost/timer/timer.hpp> or <boost/timer/progress_display.hpp>" )
      | ^~~~~~~~~~~~~~~~~~~~~~~
/usr/include/boost/timer.hpp:21:1: note: '#pragma message: This header is deprecated. Use the facilities in <boost/timer/timer.hpp> instead.'
   21 | BOOST_HEADER_DEPRECATED( "the facilities in <boost/timer/timer.hpp>" )
      | ^~~~~~~~~~~~~~~~~~~~~~~
[ 37%] Linking CXX executable fm_benchmark
[ 37%] Built target fm_benchmark
[ 43%] Building CXX object examples/CMakeFiles/full_example.dir/full_example.cpp.o
In file included from /planning/include/fast_methods/fm2/fm2.hpp:44,
                 from /planning/examples/full_example.cpp:13:
/planning/include/fast_methods/gradientdescent/gradientdescent.hpp: In static member function 'static void GradientDescent<grid_t>::apply(grid_t&, unsigned int&, GradientDescent<grid_t>::Path&, std::vector<double>&, double) [with grid_t = nDGridMap<FMCell, 3>]':
/planning/include/fast_methods/gradientdescent/gradientdescent.hpp:76:21: warning: iteration 1 invokes undefined behavior [-Waggressive-loop-optimizations]
   76 |               d_[i] = dimsize[i]*d_[i-1];
      |               ~~~~~~^~~~~~~~~~~~~~~~
/planning/include/fast_methods/gradientdescent/gradientdescent.hpp:75:32: note: within this loop
   75 |           for (size_t i = 1; i < ndims_; ++i)
      |                              ~~^~~~~~~~
[ 50%] Linking CXX executable full_example
[ 50%] Built target full_example
[ 56%] Building CXX object examples/CMakeFiles/test_fm.dir/test_fm.cpp.o
In file included from /planning/examples/test_fm.cpp:7:
/planning/include/fast_methods/ndgridmap/ndgridmap.hpp: In member function 'double nDGridMap<T, ndims>::getMaxValue() const':
/planning/include/fast_methods/ndgridmap/ndgridmap.hpp:261:28: error: 'isinf' is not a member of 'std'
  261 |                 if (!(std::isinf(c.getValue())) && c.getValue() > max)
      |                            ^~~~~
make[2]: *** [examples/CMakeFiles/test_fm.dir/build.make:76: examples/CMakeFiles/test_fm.dir/test_fm.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:215: examples/CMakeFiles/test_fm.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
````

After investigating a little I belive the use of cmath is needed since the base std does not include those functions (isinf / isnan) andcmath updates to a more recent one that has them.

ferxxp reviewed (commented):
This is unnecessary, just a sane check

ferxxp commented:
I have updated the DockerFile to the correct repository. 
I have done all changes to isinf and isnan with a successfull compilation.
I have also added at **/examples/test_fm.cpp** the library **<cmath>** due to the error I commented before.
I have laso cleansed the commits name just in case.
If there is any other suggestion just let me know

Also saludos de la UC3M y de Luis Moreno.

jvgomez commented on CMakeLists.txt:
```diff
@@ -18,6 +18,9 @@ set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "-O3 -g")
 set(CMAKE_CXX_FLAGS_RELEASE "-Ofast -fno-finite-math-only")
 set(CMAKE_CXX_FLAGS_DEBUG  "-Wall -Wno-unused-local-typedefs -g")
 
+#Flag to avoid error in compilation
```
Please remove these comments if they are no longer needed.

jvgomez replied:
Thank you for the changes! Let's clean up the CMakeLists.txt and then I believe it is good to merge.

> I have also added at /examples/test_fm.cpp the library due to the error I commented before.

What do you mean there? I do not see any changes on that file.

Saludos a vosotros también! :D

ferxxp committed:
Deleted fpermissive flag

ferxxp commented on include/fast_methods/ndgridmap/ndgridmap.hpp:
```diff
@@ -35,6 +35,7 @@
 #include <cstddef>
 #include <array>
 #include <sstream>
+#include <cmath>
```
I just copy the wrong file name. I added the **cmath** library to properly include std::isinf, else wise it cannot find it.

jvgomez commented on Dockerfile:
```diff
@@ -0,0 +1,18 @@
+FROM ubuntu:latest
```
Hi again, as I was a bit rusty on Docker, I consulted with experts and they gave me this feedback about this dockerfile:

- A "fixed" version of the base image should be used, e.g., ubuntu:22.04. latest will change with new Ubuntu releases (which will impact the versions of the installed dependencies)
- Best practice is to reduce the number of layers. In particular all the apt-get command should be one command: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#apt-get. We should also combine all the build commands (mkdir, cmake, make) into one.
- If the image were to be distributed a mutli-stage Dockerfile should be used, i.e., the output image does not require all the -dev dependencies to run => results in a smaller output image
- The last cd command is kind of useless; the CMD statement later uses all absolute paths. Also, it would be more efficient to use WORKDIR in this case (does not create a layer)
- The ARG command is probably incorrect. If the intention was to set an ENV var ENV should be used. In any case, it can be added to the RUN command without having to set it using ARG/ENV
- Instead of cloning the repository, COPY should be used COPY . /fast_methods/ makes more sense and it is safer.
- We should include a .dockerignore (same content as .gitignore) to exclude the build directory, etc. from being passed in as part of the build context (otherwise, the mkdir build command might actually fail when using COPY
- Also, please add a sentence in the README about the Docker support (how to build; how to use)

ferxxp replied:
Ok. I'm a noob with Docker also. I'll try my best.

ferxxp committed:
Updated Dockerfile and new .gitignore

Reduced image dependencies and simplified the process. multi-stage dockerfile will be explore shortly.

jvgomez reviewed (commented):
Thank you for the changes!!
Does this work as you expect? I do not have time to test it before accepting it.

jvgomez commented:
Ok, I am afraid I can't help much with the docker stuff. While you figure it out, perhaps it is a good idea to split this PR into 2, one for the compilation issue and another one for the Docker. I would be happy to merge the fix already.

ferxxp committed:
Docker moved to docker branch

ferxxp commented:
Ok, I will work on the Docker when I get some free time. For the moment lets just merge the fix as suggested.

jvgomez commented:
Thank you very much for fixing the compilation issues and helping maintaining the library in a good state.
